ipn/ipnlocal/local.go

```diff
@@ -2047,6 +2054,43 @@ func (b *LocalBackend) updateFilterLocked(netMap *netmap.NetworkMap, prefs ipn.P
 	}
 }
 
+// captivePortalWarnable is a Warnable which is set to an unhealthy state when a captive portal is detected.
+var captivePortalWarnable = health.Register(&health.Warnable{
+	Code:  "captive-portal-detected",
+	Title: "Captive portal detected",
+	// High severity, because captive portals blocks all traffic and require user intervention.
+	Severity:            health.SeverityHigh,
+	Text:                health.StaticMessage("This Wi-Fi network requires you to log in manually."),
+	ImpactsConnectivity: true,
+})
+
+// performCaptiveDetection checks if captive portal detection is enabled via controlknob. If so, it runs
+// the detection and updates the Warnable accordingly.
+func (b *LocalBackend) performCaptiveDetection() {
+	captiveDetectionEnabledByControlKnob := b.ControlKnobs().CaptivePortalDetection.Load()
+	if !captiveDetectionEnabledByControlKnob {
+		return
+	}
+
+	d := captivedetection.NewDetector()
+	var dm *tailcfg.DERPMap = nil
+	if b.netMap != nil {
+		dm = b.netMap.DERPMap
+	}
```

Comment:
You need to hold `b.mu` to access `b.netMap`. Also, stylistically, you don't need to initialize the `dm` variable to `nil`; that's the default ("zero value") in Go.

Reply:
Done.